### PR TITLE
test(e2e, playwright): challenge-title.tsx

### PR DESCRIPTION
**Pull request by EggSaled**
Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Refers: #51705

### CODE REVIEW

**Comment by EggSaled** [comment]
camperbot cant code

**Comment by EggSaled** [comment]
I am not quite sure if I should also be testing for `<GreenPass />` or for the code block below.
```jsx
      {translationPending && (
        <>
          <Link
            className='title-translation-cta'
            to={i18next.t('links:help-translate-link-url')}
          >
            {i18next.t('misc.translation-pending')}
          </Link>
        </>
      )}

```

If those are additional cases that need testing please let me know.

**Review by huyenltnguyen — COMMENTED** [review]
> I am not quite sure if I should also be testing for `<GreenPass />`

We should also have an assertion to ensure that the check mark shows up on the screen. GreenPass has a label of "Passed", so we should be able to query it using the `getByLabel` locator: https://playwright.dev/docs/api/class-framelocator#frame-locator-get-by-label.

---
As for the `translationPending` rendering condition, ideally we should add a test case for it, though I'm not sure if it's trivial to set up the test (looks like the condition involves environment variables). You could give it a try and let us know your findings or if you need help from us.

File: e2e/challenge-title.spec.ts
Comment (EggSaled) [reply e2e/challenge-title.spec.ts:26]: Should I remove the search for both `challenge-title-wrap` and `challenge-title`? Or should I still keep the the test for the `challenge-title-wrap`?

**Comment by EggSaled** [comment]
I've added the test for `GreenPass`. However, I made a change to the challenge being used to test out when `GreenPass` should be visible, since the original challenge required linking a Microsoft account to complete the challenge. Also, there was a weird issue I encountered while running the tests where the final check for `GreenPass` would sometimes timeout.

----

As for the `translationPending` condition, I was having trouble figuring out how to come up with an adequate test. The current thought process I have at the moment is this:

- Use `i18next` to look for the "Help us Translate" button, if it exists
- Since untranslated text falls back to English, verify if there isn't a translated version of the title

File: e2e/challenge-title.spec.ts
Comment (EggSaled) [reply e2e/challenge-title.spec.ts:35]: Playwright would raise errors about either redirecting while loading a different page or return `NS_BINDING_ABORTED`. The `page.waitForLoadState()` was an attempt to quell those errors. I'll remove those if they haven't been happening on your end.

----

The redirect was meant to check if `GreenPass` would render on the challenge page itself, but I didn't notice that there is an additional `GreenPass` that renders on the courses page as well. Should I just check for `GreenPass` on the courses page or keep checking on the challenge page itself?

File: e2e/challenge-title.spec.ts
Comment (EggSaled) [reply e2e/challenge-title.spec.ts:18]: If I recall correctly, I did make an initial attempt to query the element using `getByRole`, but encountered some issues with it. I used a different `name` parameter though, so i'll try out this suggestion and let you know if I encounter similar issues.

File: e2e/challenge-title.spec.ts
Comment (EggSaled) [inline e2e/challenge-title.spec.ts:44]: The `goto` raises an `NS_BINDING_ABORTED` error (or a navigation interruption error) for Firefox, Webkit, and Mobile Safari. The `waitForLoadState` was an attempt at resolving these errors for the browsers mentioned, but the problem still persists.

File: e2e/challenge-title.spec.ts
Comment (EggSaled) [inline e2e/challenge-title.spec.ts:100]: Same `NS_BINDING_ABORTED` error raised here as well.

**Review by huyenltnguyen — APPROVED** [review]
@EggSaled I went ahead and made some changes to the PR to shorten the review loop: [ac11e43](https://github.com/freeCodeCamp/freeCodeCamp/pull/51818/commits/ac11e431c7cf38562e92ba376c2b75ba791a8128). 

Thank you for your contribution and your patience! This is one heck of a challenge, so we really appreciate your effort here.

---
What I changed:
1. Removed the `afterEach` block as Playwright already handles this for us
2. Replaced `getByText()` with `getByRole()` to that we can also cover the accessibility front
3. Made the authentication and authentication reset more explicit with the `storageState` override
4. Replaced `waitForLoadState()` with `waitForURL()` in the first navigation (the one where the user completes the challenge and goes to the curriculum list)
5. Skipped the signed-in test case on Webkit

---
I think you're most interested in (4), so here is the explanation (from what I understand).

There is an issue thread regarding the `NS_BINDING_ABORTED` error: https://github.com/microsoft/playwright/issues/20749.

What happened was that when the user goes to the next challenge (or the curriculum list in this case) then goes back to the original challenge, there are 2 navigations the browser needs to make. The second navigation comes in before Firefox finishes the first one, and it aborts the first one as it sees the second. The `NS_BINDING_ABORTED` error is just an indication of the event, but it is thrown as an exception and fails the test.

In the Playwright issue thread, the maintainer said that:
> You can wait for the redirect to finish before starting next navigation

I think this approach makes sense, as it matches the actual workflow: the user is sent to the curriculum page, they would need to wait for the page to load, find and click on the challenge themselves. We are using `page.goto()` in the test as a shortcut, which is okay, but I think we still need to ensure that the curriculum page (the first navigation) completes properly.

And in the documentation of [`waitForLoadState()`](https://playwright.dev/docs/api/class-page#page-wait-for-load-state), there is a note that:
> The navigation must have been committed when this method is called.

I read "committed" as "completed", so I think `waitForURL()` is more appropriate for the first navigation, and we can use `waitForLoadState()` for the second.

This was the test result after the change:

Screenshot 2023-11-06 at 01 47 16

---

The failure on Webkit doesn't seem to be related to the `NS_BINDING_ABORTED` error but rather the authentication. This is a known issue and we are going to investigate separately, so I temporarily skipped the test on this browser to unblock the PR.

Here is the error (signed-in user should see the check mark, but the test disagreed):

Screenshot 2023-11-06 at 02 43 57

**Review by moT01 — APPROVED** [review]
LGTM @EggSaled 🎉 Thanks for contributing to freeCodeCamp 🎉

**Comment by EggSaled** [comment]
Thank you Huyen for helping out on this PR! I am glad to have taken up this task, especially since I was able to learn a lot about how Playwright works.

----

> I read "committed" as "completed", so I think `waitForURL()` is more appropriate for the first navigation, and we can use `waitForLoadState()` for the second.

When you put it that way, it makes a lot more sense as to why I wasn't able to fix the `NS_BINDING_ABORTED` error. If I am understanding this part correctly, Playwright treats page navigation as transactional, and the approach I was taking wasn't taking that information into consideration. At the very least, I am able to better understand when to use `waitForURL()` and `waitForLoadState()`.

----

Once again, thank you Huyen! I certainly learned quite a lot on this PR, and I am glad to have taken up this task! :smile:

**Pull request merged**
